src/cockpit/channels/filesystem.py

```diff
@@ -188,12 +188,23 @@ def do_done(self):
             if self._tag and self._tag != tag_from_path(self._path):
                 raise ChannelError('change-conflict')
 
+            # if the file exists already, record its permissions
+            try:
+                stat = os.stat(self._path)
+            except (FileNotFoundError, PermissionError):
+                stat = None
+
             try:
                 os.rename(self._temppath, self._path)
             except OSError:
                 # ensure to not leave the temp file behind
                 self.unlink_temppath()
                 raise
+            # restore original permissions
+            if stat is not None:
+                os.chmod(self._path, stat.st_mode)
+                os.chown(self._path, stat.st_uid, stat.st_gid)
```

Comment:
Hmm, now that I look at this, I think we should do this on `temppath` instead, to get a better failure mode if the permission restoration fails.

Reply:
Done.